ldms/src/sampler/zfs_leafvdevs/zfs_leafvdevs.c

```diff
@@ -0,0 +1,496 @@
+/* -*- c-basic-offset: 8 -*- */
+/* Copyright 2022 Lawrence Livermore National Security, LLC
+ * See the top-level COPYING file for details.
+ *
+ * SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause)
+ */
+
+#define _GNU_SOURCE
+#include <stdlib.h>
+#include <ctype.h>
+#include <glob.h>
+#include <string.h>
+#include <getopt.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <inttypes.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <libzfs.h>
+#include <libzutil.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include "ldms.h"
+#include "ldmsd.h"
+#include "config.h"
+#include "sampler_base.h"
+#include "libzfs.h"
+#include <stddef.h>
+
+#define SAMP "zfs_leafvdevs"
+#ifndef PATH_MAX
+#define PATH_MAX 2048
+#endif
+
+#ifndef ARRAY_LEN
+#define ARRAY_LEN(a) (sizeof(a) / sizeof(*a))
+#endif
+
+/* Function prototypes */
+
+int get_leaf_stats(zpool_handle_t *, void *);
+
+typedef int (*stat_printer_f) (nvlist_t *, const char *, zpool_handle_t *);
+
+static int vdevs_count(zpool_handle_t *, void *);
+
+static ldmsd_msg_log_f log_fn;
+static base_data_t sampler_base;
+static libzfs_handle_t *g_zfs;
+
+static struct {
+	int vdev_list_idx;
+	int vdev_rec_idx;
+} index_store;
+
+static ldms_mval_t list_handle;
+
+/* metric templates for a virtual device */
+static struct ldms_metric_template_s zfs_leafvdevs[] = {
+	{"zpoolname", 0, LDMS_V_CHAR_ARRAY, "", ZFS_MAX_DATASET_NAME_LEN},
+	{"state", 0, LDMS_V_CHAR_ARRAY, "", ZFS_MAX_DATASET_NAME_LEN},
+	{"vdevname", 0, LDMS_V_CHAR_ARRAY, "", ZFS_MAX_DATASET_NAME_LEN},
+	{"vdevpath", 0, LDMS_V_CHAR_ARRAY, "", ZFS_MAX_DATASET_NAME_LEN},
+	{"vdevguid", 0, LDMS_V_CHAR_ARRAY, "", ZFS_MAX_DATASET_NAME_LEN},
+	{"alloc", 0, LDMS_V_U64, "", 1},
+	{"free", 0, LDMS_V_U64, "", 1},
+	{"size", 0, LDMS_V_U64, "", 1},
+	{"read_bytes", 0, LDMS_V_U64, "", 1},
+	{"read_errors", 0, LDMS_V_U64, "", 1},
+	{"read_ops", 0, LDMS_V_U64, "", 1},
+	{"write_bytes", 0, LDMS_V_U64, "", 1},
+	{"write_errors", 0, LDMS_V_U64, "", 1},
+	{"write_ops", 0, LDMS_V_U64, "", 1},
+	{"checksum_errors", 0, LDMS_V_U64, "", 1},
+	{"fragmentation", 0, LDMS_V_U64, "", 1},
+	{"init_errors", 0, LDMS_V_U64, "", 1},
```

Comment:
As you did for `zfs_topvdevs.c`, "trim_errors" and "slow_ios" should be added here.

Reply:
Trim_errors and slow_io have been added.